R/utils.R

```diff
@@ -0,0 +1,74 @@
+#' Wrapper function to help generate a packrat lockfile
+#'
+#' @param ...      the arguments to pass to gen_runtime_description
+#' @param .workdir directory to pass to install_from_desc
+#' @export
+gen_packrat_lockfile <- function(..., .workdir = file.path( getwd(),"work")){
+  
+  runtime_desc <- pkglock::gen_runtime_description(...)
+  install <- pkglock::install_from_desc(runtime_desc, .dir = .workdir)
+  lockfile <- install$snapshot$lockfile
+  
+  return (lockfile)
+}
+
+#' Scan a library folder and build a package list
+#' @param .libdir the library directory to scan
+#' @export
+build_package_list_from_library_dir <- function(.libdir){
+  
+  pkgs <- list();
+  for( file in list.files(.libdir)){
+    version <-  packageDescription(file, lib.loc = .libdir)$Version
+    pkgs[file] <- version
+  }
+  
+  return (pkgs)
+}
+
+
+#' Build a string suitable for passing to gen_runtime_description
+#' @param .pkglist the output from build_package_list_from_library_dir
+#' @param .addversion  TRUE to lock package to version, FALSE to get latest
+#' @export
+gen_pkg_desc_from_package_list <- function(.pkglist, .addversion=FALSE){
+  out <- ''
+  
+  pkgnames <- names(.pkglist)
+  add <- FALSE
+  for( name in pkgnames){
+    out <- sprintf(
+      "%s%s'%s%s'",
+      out,
+      if(add) ",\n\t  " else "",
+      name,
+      if(.addversion) paste0(" (==",.pkglist[name],")") else ""
+    )
+    add <- TRUE
+  }
+  
+  return (paste0("pkgs <- c(",out,")"))
+}

```

Comment:
after seeing what you were doing with this in the TEST section below, I think it makes more sense to just keep it simple and return as a character vector. You don't need to format it with spaces etc and call it pkgs unless you are using this to autogen code for someone to paste elsewhere. If you want to pass directly to gen_pkg_desc

Reply:
This was just something I was playing with to auto-generate some code so that the actual pkglock gen request can be committed to git as the source that generate the packrat lockfile.  Mostly this was to simplify getting a starting point.  But, I suspect this might be a bad idea, because this is making every installed package a "TOP" level package request that will then get all of their imports,suggests, etc... when installing which will blow up the list of installed packages even more.

Reply:
two things now that I've looked at this:

1) I think having such a function is a good idea, given the desire to snapshot a given directory

2) This is a bit circuitous as essentially as the purpose of the gen_runtime_desc is doing a bunch of extra work to install a set of packages that don't yet exist on the system. 

Instead I think a better function to cover such a behavior would be `gen_lockfile_from_dir(pkgs = ..., .libdir = ...)` where the top level packages can be declared, then only the tree of requirements from those packages can be pruned from an existing library install.